rust/arrow/src/array/union.rs

```diff
@@ -0,0 +1,1174 @@
+    pub fn value(&self, index: usize) -> ArrayRef {
+        let type_id = self.type_id(self.offset() + index);
+        let value_offset = self.value_offset(self.offset() + index) as usize;
+        let child_data = self.boxed_fields[type_id as usize].clone();
+        child_data.slice(value_offset, 1)
+    }
+
+    /// Returns the names of the types in the union.
+    pub fn type_names(&self) -> Vec<&str> {
+        match self.data.data_type() {
+            DataType::Union(fields) => fields
+                .iter()
+                .map(|f| f.name().as_str())
+                .collect::<Vec<&str>>(),
+            _ => unreachable!("Union array's data type is not a union!"),
+        }
+    }
+
+    /// Returns whether the `UnionArray` is dense (or sparse if `false`).
+    fn is_dense(&self) -> bool {
+        self.data().buffers().len() == 2
+    }
+}
+
+impl From<ArrayDataRef> for UnionArray {
+    fn from(data: ArrayDataRef) -> Self {
+        let mut boxed_fields = vec![];
+        for cd in data.child_data() {
+            boxed_fields.push(make_array(cd.clone()));
+        }
+        Self { data, boxed_fields }
+    }
+}
+
+impl Array for UnionArray {
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
+    }
+}
+
+impl fmt::Debug for UnionArray {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let header = if self.is_dense() {
+            "UnionArray(Dense)\n[\n"
+        } else {
+            "UnionArray(Sparse)\n[\n"
+        };
+        write!(f, "{}", header)?;
+
+        write!(f, "-- type id buffer:\n")?;
+        write!(f, "{:?}\n", self.data().buffers()[0])?;
+
+        if self.is_dense() {
+            write!(f, "-- offsets buffer:\n")?;
+            write!(f, "{:?}\n", self.data().buffers()[1])?;
+        }
+
+        for (child_index, name) in self.type_names().iter().enumerate() {
+            let column = &self.boxed_fields[child_index];
+            write!(
+                f,
+                "-- child {}: \"{}\" ({:?})\n",
+                child_index,
+                *name,
+                column.data_type()
+            )?;
+            fmt::Debug::fmt(column, f)?;
+            write!(f, "\n")?;
+        }
+        write!(f, "]")
+    }
+}
+
+/// `FieldData` is a helper struct to track the state of the fields in the `UnionBuilder`.
+struct FieldData {
+    /// The type id for this field
+    type_id: i8,
+    /// The Arrow data type represented in the `values_buffer`, which is untyped
+    data_type: DataType,
+    /// A buffer containing the values for this field in raw bytes
+    values_buffer: Option<MutableBuffer>,
+    ///  The number of array slots represented by the buffer
+    slots: usize,
+    /// The number of null array slots in this child array
+    null_count: usize,
+    /// A builder for the bitmap if required
```

Comment:
When is this required vs not?

Reply:
It's required for sparse unions and not for dense, I'll update the doc comment.